packages/cli/src/util/env/refresh-oidc-token.ts

```diff
@@ -0,0 +1,115 @@
+import { decodeJwt } from 'jose';
+import ms from 'ms';
+import { performance } from 'perf_hooks';
+import output from '../../output-manager';
+import type Client from '../../util/client';
+import {
+  type EnvRecordsSource,
+  pullEnvRecords,
+} from '../../util/env/get-env-records';
+import sleep from '../../util/sleep';
+import { VERCEL_OIDC_TOKEN } from './constants';
+
+const REFRESH_BEFORE_EXPIRY_MS = getMs(
+  ms('15m'),
+  process.env.REFRESH_VERCEL_OIDC_TOKEN_BEFORE_EXPIRY_MS
+);
+
+const THROTTLE_MS = getMs(
+  ms('1m'),
+  process.env.REFRESH_VERCEL_OIDC_TOKEN_THROTTLE_MS
+);
+
+function getMs(defaultValue: number, overrideValue?: string): number {
+  if (overrideValue) {
+    const result = ms(overrideValue);
+    if (Number.isFinite(result) && result > 0) return result;
+  }
+  return defaultValue;
+}
+
+export async function* refreshOidcToken(
+  client: Client,
+  projectId: string,
+  envValues: Record<string, string>,
+  source: EnvRecordsSource
+): AsyncGenerator<string> {
+  let lastPulledEnvAt = clock();
+  let refreshCount = 0;
+
+  while (true) {
+    // If the user has OIDC disabled, do nothing.
+    const oidcToken = envValues[VERCEL_OIDC_TOKEN];
+    if (!oidcToken) {
+      output.debug(`${VERCEL_OIDC_TOKEN} is absent; disabling refreshes`);
+      return;
+    }
+
+    // Otherwise, extract the "exp" claim.
+    const now = clock();
+    const { exp } = decodeJwt(oidcToken);
+    const expiresAfterMs = exp !== undefined ? exp * 1000 - now : undefined;
```

Comment:
nit: would name `Ms` as `Millis` and `S` as `Secs` for better readability throughout PR.